### PR TITLE
Fix cross version test python version inference

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/12832?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12832/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12832
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Fix cross version test python version inference

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [X] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

File: dev/set_matrix.py [hunk 1]
```diff
@@ -261,7 +261,7 @@ def get_requires_python(package: str, version: str) -> str:
         ),
         None,
     )
-    candidates = ("3.8", "3.9")
+    candidates = ("3.8", "3.9", "3.10")
```
Comment (harupy) [inline dev/set_matrix.py:264]: thought this would work, but it doesn't. Can you check the `requires_python` value of xgboost?

Comment (WeichenXu123) [reply dev/set_matrix.py:264]: I think `get_requires_python` only works for released version ?

File: mlflow/ml-package-versions.yml [hunk 2]
```diff
@@ -168,6 +169,9 @@ xgboost:
     requirements:
       ">= 0.0.0": ["scikit-learn", "matplotlib"]
       "< 1.6": ["pandas<2"]
+      "== dev": ["setuptools==59.8.0"]
```
Comment (WeichenXu123) [inline mlflow/ml-package-versions.yml:172]: this is for addressing a setuptools bug: https://github.com/mlflow/mlflow/actions/runs/10173505108/job/28137876205?pr=12832#step:12:274

File: .github/actions/setup-python/action.yml [hunk 3]
```diff
@@ -32,8 +32,10 @@ runs:
           fi
         elif [[ "$python_version" == "3.9" ]]; then
           python_version="3.9.18"
+        elif [[ "$python_version" == "3.10" ]]; then
+          python_version="3.10.14"
         else
-          echo "Invalid python version: '$python_version'. Must be '3.8'"
+          echo "Invalid python version: '$python_version'. Must be '3.8' or '3.9' or '3.10'"
```
Comment (harupy) [inline .github/actions/setup-python/action.yml:38]:
```suggestion
          echo "Invalid python version: '$python_version'. Must be '3.8', '3.9', or '3.10'"
```

File: dev/set_matrix.py (hunk 1 above)
Comment (harupy) [reply dev/set_matrix.py:264]: ah you're right

File: mlflow/ml-package-versions.yml (hunk 2 above)
Comment (harupy) [reply mlflow/ml-package-versions.yml:172]: this issue is reported here: https://github.com/scikit-learn/scikit-learn/issues/26992

can you try importing `setuptools` first before `mlflow.sklearn.autolog`?

Comment (WeichenXu123) [reply mlflow/ml-package-versions.yml]: Tested, this way " importing setuptools first before mlflow.sklearn.autolog" does not address it: https://github.com/mlflow/mlflow/actions/runs/10174566035/job/28140563273#step:12:132

Comment (harupy) [reply mlflow/ml-package-versions.yml]: @WeichenXu123 thanks! can you reproduce this on your machine?

Comment (WeichenXu123) [reply mlflow/ml-package-versions.yml]: No. I can't repro it locally.

**Comment by harupy** [comment]
@WeichenXu123 I was able to reproduce the error:

Code:

```
# a.py
import mlflow
from sklearn.cluster import KMeans
from sklearn.datasets import load_iris
import sys

model = KMeans()
iris = load_iris()
X = iris.data[:, :2]
y = iris.target

with mlflow.start_run():
    model.fit(X, y)
    mlflow.sklearn.log_model(model, "model")
    print("distutils" in sys.modules)

mlflow.sklearn.autolog()
```

Run:

```
docker run -it -v $(pwd)/a.py:/a.py python:3.10 bash
```

**Comment by WeichenXu123** [comment]
Thanks! I repro it too. the call stack that importing distutils is:

```
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/mlflow/sklearn/__init__.py(412)log_model()
-> return Model.log(
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/mlflow/models/model.py(698)log()
-> flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/mlflow/sklearn/__init__.py(330)save_model()
-> _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/mlflow/utils/environment.py(90)current()
-> build_dependencies=cls.get_current_build_dependencies(),
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/mlflow/utils/environment.py(105)get_current_build_dependencies()
-> version = _PythonEnv._get_package_version(package)
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/mlflow/utils/environment.py(97)_get_package_version()
-> return __import__(package_name).__version__
  <frozen importlib._bootstrap>(1027)_find_and_load()
  <frozen importlib._bootstrap>(1006)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(688)_load_unlocked()
  <frozen importlib._bootstrap_external>(883)exec_module()
  <frozen importlib._bootstrap>(241)_call_with_frames_removed()
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/setuptools/__init__.py(8)<module>()
-> import _distutils_hack.override  # noqa: F401
  <frozen importlib._bootstrap>(1027)_find_and_load()
  <frozen importlib._bootstrap>(1006)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(688)_load_unlocked()
  <frozen importlib._bootstrap_external>(883)exec_module()
  <frozen importlib._bootstrap>(241)_call_with_frames_removed()
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/_distutils_hack/override.py(1)<module>()
-> __import__('_distutils_hack').do_override()
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/_distutils_hack/__init__.py(77)do_override()
-> ensure_local_distutils()
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/site-packages/_distutils_hack/__init__.py(60)ensure_local_distutils()
-> importlib.import_module('distutils')
  /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/importlib/__init__.py(126)import_module()
-> return _bootstrap._gcd_import(name[level:], package, level)
  <frozen importlib._bootstrap>(1050)_gcd_import()
  <frozen importlib._bootstrap>(1027)_find_and_load()
  <frozen importlib._bootstrap>(1006)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(688)_load_unlocked()
  <frozen importlib._bootstrap_external>(883)exec_module()
  <frozen importlib._bootstrap>(241)_call_with_frames_removed()
> /home/weichen.xu/miniconda3/envs/py310/lib/python3.10/distutils/__init__.py(22)<module>()->None
```

**Comment by harupy** [comment]
Importing `setuptools` before `log_model` seems to work.

**Comment by WeichenXu123** [comment]
> return __import__(package_name).__version__

yes , this is strange.

solely run `__import__('setuptools').__version__`  works.
but the same call inside `log_model` raise error.

**Comment by WeichenXu123** [comment]
The successful run invokes `python3.10/site-packages/setuptools/_distutils/__init__.py`,
and the failed run invokes `python3.10/distutils/__init__.py`

the condition that triggers the failed run is still not clear.

**Comment by WeichenXu123** [comment]
I think it relates to importing order issue , when `python3.10/site-packages/setuptools/_distutils` and `python3.10/distutils/__init__.py` coexists, importing which one is non-deterministic. but I am not sure how the importing finding route works internally.

**Comment by harupy** [comment]
@WeichenXu123 What imports `distutils`?

**Comment by WeichenXu123** [comment]
> @WeichenXu123 What imports `distutils`?

`__import__('setuptools').__version__` triggers importing `distutils` @harupy

**Comment by harupy** [comment]
@WeichenXu123 I figured out a solution. Filed #12847.

**Pull request closed without merging**